crates/ruff_linter/src/rules/refurb/rules/slice_to_remove_prefix_or_suffix.rs

```diff
@@ -0,0 +1,479 @@
+///   - `slice` has no upper bound in the case of a prefix,
+///     and no lower bound in the case of a suffix
+///
+/// If these conditions are satisfied, the function
+/// returns the corresponding `RemoveAffixData` object;
+/// otherwise it returns `None`.
+fn affix_removal_data<'a>(
+    value: &'a ast::Expr,
+    test: &'a ast::Expr,
+    else_or_target: &'a ast::Expr,
+    slice: &'a ast::Expr,
+) -> Option<RemoveAffixData<'a>> {
+    let compr_value = ast::comparable::ComparableExpr::from(value);
+    let compr_else_or_target = ast::comparable::ComparableExpr::from(else_or_target);
+    if compr_value != compr_else_or_target {
+        return None;
+    }
+    let slice = slice.as_slice_expr()?;
+    let compr_test_expr = ast::comparable::ComparableExpr::from(
+        &test.as_call_expr()?.func.as_attribute_expr()?.value,
+    );
+    let func_name = test
+        .as_call_expr()?
+        .func
+        .as_attribute_expr()?
+        .attr
+        .id
+        .as_str();
+
+    let func_args = &test.as_call_expr()?.arguments.args;
+
+    let [affix] = func_args.as_ref() else {
+        return None;
+    };
+    if compr_value != compr_test_expr || compr_test_expr != compr_else_or_target {
+        return None;
+    }
+    let (affix_kind, bound) = match func_name {
+        "startswith" if slice.upper.is_none() => (AffixKind::StartsWith, slice.lower.as_ref()?),
+        "endswith" if slice.lower.is_none() => (AffixKind::EndsWith, slice.upper.as_ref()?),
+        _ => return None,
+    };
+    Some(RemoveAffixData {
+        text: value,
+        bound,
+        affix_query: AffixQuery {
+            kind: affix_kind,
+            affix,
+        },
+    })
+}
+
+/// Tests whether the slice of the given string actually removes the
+/// detected affix.
+///
+/// For example, in the situation
+///
+/// ```python
+///  text[:bound] if text.endswith(suffix) else text
+/// ```
+///
+/// This function verifies that `bound == -len(suffix)` in two cases:
+///   - `suffix` is a string literal and `bound` is a number literal
+///   - `suffix` is an expression and `bound` is
+///     exactly `-len(suffix)` (as AST nodes, prior to evaluation.)
+fn affix_matches_slice_bound(data: &RemoveAffixData, semantic: &SemanticModel) -> bool {
+    let RemoveAffixData {
+        text: _,
+        bound,
+        affix_query: AffixQuery { kind, affix },
+    } = *data;
+
+    match (kind, bound, affix) {
+        (
+            AffixKind::StartsWith,
+            ast::Expr::NumberLiteral(ast::ExprNumberLiteral {
+                value: num,
+                range: _,
+            }),
+            ast::Expr::StringLiteral(ast::ExprStringLiteral {
+                range: _,
+                value: string_val,
+            }),
+        ) => num.as_int().is_some_and(|x| {
+            // Only support prefix removal for size at most `u32::MAX`
+            u32::try_from(string_val.len()).is_ok_and(|length| x == &ast::Int::from(length))
+        }),
```

Comment:
I think this is what @MichaReiser was suggesting earlier (I checked locally and this version seems to compile! You need to add an import of the `ruff_text_size::TextLen` trait though)

```suggestion
        ) => num
            .as_int()
            .and_then(ast::Int::as_u32) // Only support prefix removal for size at most `u32::MAX`
            .is_some_and(|x| x == string_val.to_str().text_len().to_u32()),
```